synapseclient/models/mixins/asynchronous_job.py

```diff
@@ -0,0 +1,372 @@
+import asyncio
+import json
+import time
+from dataclasses import dataclass
+from enum import Enum
+from typing import Any, Dict, Optional
+
+from synapseclient import Synapse
+from synapseclient.core.constants.concrete_types import AGENT_CHAT_REQUEST
+from synapseclient.core.exceptions import SynapseError, SynapseTimeoutError
+
+ASYNC_JOB_URIS = {
+    AGENT_CHAT_REQUEST: "/agent/chat/async",
+}
+
+
+class AsynchronousCommunicator:
+    """Mixin to handle communication with the Synapse Asynchronous Job service."""
+
+    def to_synapse_request(self) -> None:
+        """Converts the request to a request expected of the Synapse REST API."""
+        raise NotImplementedError("to_synapse_request must be implemented.")
+
+    def fill_from_dict(
+        self, synapse_response: Dict[str, str]
+    ) -> "AsynchronousCommunicator":
+        """
+        Converts a response from the REST API into this dataclass.
+
+        Arguments:
+            synapse_response: The response from the REST API.
+
+        Returns:
+            An instance of this class.
+        """
+        raise NotImplementedError("fill_from_dict must be implemented.")
+
+    async def _post_exchange_async(
+        self, synapse_client: Optional[Synapse] = None, **kwargs
+    ) -> None:
+        """Any additional logic to run after the exchange with Synapse.
+
+        Arguments:
+            synapse_client: The Synapse client to use for the request.
+            **kwargs: Additional arguments to pass to the request.
+        """
+        pass
+
+    async def send_job_and_wait_async(
+        self,
+        post_exchange_args: Optional[Dict[str, Any]] = None,
+        *,
+        synapse_client: Optional[Synapse] = None,
+    ) -> "AsynchronousCommunicator":
+        """Send the job to the Asynchronous Job service and wait for it to complete.
+
+        Arguments:
+            post_exchange_args: Additional arguments to pass to the request.
```

Comment:
Could you update what I wrote here to be a little more verbose on how this is intended to be used? Maybe an example too?

Could you also update the language for the `synapse_client` docstring to match the other areas of the code?